src/Microsoft.Azure.SignalR/HubHost/ServiceLifetimeManager.cs

```diff
@@ -103,6 +112,85 @@ public override async Task SendConnectionAsync(string connectionId, string metho
+            if (IsInvalidArgument(connectionId))
+            {
+                throw new ArgumentNullException(nameof(connectionId));
+            }
+
+            if (IsInvalidArgument(methodName))
+            {
+                throw new ArgumentNullException(nameof(methodName));
+            }
+
+            var invocationId = _clientInvocationManager.Caller.GenerateInvocationId(connectionId);
+            var message = AppendMessageTracingId(new ClientInvocationMessage(invocationId, connectionId, _callerId, SerializeAllProtocols(methodName, args, invocationId)));
+            await WriteAsync(message);
+            var task = _clientInvocationManager.Caller.AddInvocation<T>(connectionId, invocationId, cancellationToken);
+
+            // Exception handling follows https://source.dot.net/#Microsoft.AspNetCore.SignalR.Core/DefaultHubLifetimeManager.cs,349
+            try
+            {
+                return await task;
+            }
+            catch
+            {
+                _clientInvocationManager.Caller.RemoveInvocation(invocationId);
+                throw;
+            }
+        }
+
+        public override async Task SetConnectionResultAsync(string connectionId, CompletionMessage result)
+        {
+            if (IsInvalidArgument(connectionId))
+            {
+                throw new ArgumentException(NullOrEmptyStringErrorMessage, nameof(connectionId));
+            }
+            if (_clientConnectionManager.ClientConnections.TryGetValue(connectionId, out var clientConnectionContext))
+            {
+                // Determine which manager (Caller / Router) the `result` belongs to.
+                // `TryCompletionResult` returns false when the corresponding invocation is not existing.
+                IClientResultsManager clientResultsManager = null;
+                if (_clientInvocationManager.Caller.TryCompleteResult(connectionId, result))
+                {
+                    clientResultsManager = _clientInvocationManager.Caller;
+                }
+                if (_clientInvocationManager.Router.TryCompleteResult(connectionId, result))
+                {
+                    clientResultsManager = _clientInvocationManager.Router;
+                }
+
+                // Block unknown `results` which belongs to neither Caller nor Router
+                if (clientResultsManager != null)
+                {
+                    var protocol = clientConnectionContext.Protocol;
+                    // For router server, it should send a ClientCompletionMessage with accurate payload content, which is necessary for the caller server.
+                    // For caller server, the only purpose of sending ClientCompletionMessage is to inform service to cleanup the invocation, which means only InvocationId and ConnectionId make sense. To avoid serialization for useless payload, we assign payload with empty bytes.
+                    var payload = clientResultsManager == _clientInvocationManager.Router
```

Comment:
How about move payload assignment along with the check during L156~L162. So you don't have to do another manager type check here.

Reply:
moved